WordPress/wordpress-playground
# [Patch WordPress] Ensure block editor iframe is controlled by a service worker

## What is this PR doing?

Patches the block editor to use a special ControlledIframe component instead of a regular HTML "iframe" element. The goal is to make the iframe use a plain HTTP URL instead of srcDoc, blob URL and other variations.

Normally, the patch applied here would be a huge maintenance burden over time. However, @ellatrix explores fixing the issue upstream [in the Gutenberg repo](https://github.com/WordPress/wordpress-playground/issues/646). Once her PR is merged, the patch here will only be needed for a known and limited set of WordPress and Gutenberg versions and will not require ongoing reconciliation with new WP/GB releases.

Fixes #646

## What problem is it solving?

In Playground, the editor iframe needs to be controlled by Playground's service worker so it can serve CSS and other static assets. Otherwise all the requests originating in that iframe will yield 404s.

However, different WordPress versions use a variety of iframe techniques that result in a non-controlled iframe:

* 6.3 uses a binary blob URL and the frame isn't controlled by a service worker
* <= 6.2 uses srcdoc had a null origin and the frame isn't controlled by a service worker
* Other dynamic techniques, such as using a data URL, also fail to produce a controlled iframe

HTTP URL src like src="/doc.html" seems to be the only way to create a controlled iframe.

And so, this commit ensures that the block editor iframe uses a plain HTTP URL regardless of the WordPress version.

Once https://github.com/WordPress/gutenberg/pull/55152 lands, this will just work in WordPress 6.4 and new Gutenberg releases.

## Testing Instructions

Run `npm run dev`

Then, confirm the inserter is nicely styled and there are no CSS-related 404s in the network tools.

Test the following editors:

* Post editor http://localhost:5400/website-server/?url=/wp-admin/post-new.php
* Site editor http://localhost:5400/website-server/?url=/wp-admin/site-editor.php
* For all supported WordPress versions
* With and without the Gutenberg plugin (`&plugin=gutenberg`)


## Related

* https://bugs.chromium.org/p/chromium/issues/detail?id=880768
* https://bugzilla.mozilla.org/show_bug.cgi?id=1293277
* https://github.com/w3c/ServiceWorker/issues/765
* https://github.com/WordPress/wordpress-playground/issues/42
* b7ca737
